File: Algo-1/week3/5-Bandwidth-Manager/bandwidth_manager.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>
#include <map>

using namespace std;

typedef pair<pair<string, int>, string> PPSIS;

// priority of protocols
map<string, int> Protocols = {
	{ "ICMP", 1 },
	{ "UDP", 2 },
	{ "RTM", 3 },
	{ "IGMP", 4 },
	{ "DNS", 5 },
	{ "TCP", 6 }
};

class BandwidthManager {
private:
	struct CompareProtocols {
		bool operator() (const PPSIS &a, const PPSIS &b) {
			if (Protocols[a.first.first] == Protocols[b.first.first]) {
				return a.first.second > b.first.second;
			}

			return Protocols[a.first.first] > Protocols[b.first.first];
		}
	};

	priority_queue<PPSIS, vector<PPSIS>, CompareProtocols> PriorityQ;

public:
	//receives a packet with specified protocol and payload
	void rcv(string protocol, string payload) {
		static int counter = 0;
		PriorityQ.push(make_pair(make_pair(protocol, counter++), payload));
	}

	//returns the payload of the packet which should be send
	string send() {
		if (PriorityQ.empty()) {
			return "Nothing to send!";
		}

		string result = PriorityQ.top().second;
		PriorityQ.pop();
		
		return result;
	}
};


int main() {

	cin.sync_with_stdio(false);
	cin.tie(0);

	int n;
	cin >> n;

	BandwidthManager test;

	for (int i = 0; i < n; i++) {
		string command, Protocol, Payload;
		cin >> command;
		
		if (command == "rcv") {
			cin >> Protocol;
			cin >> Payload;
			
			test.rcv(Protocol, Payload);
		}
		else if (command == "send") {
			cout << test.send() << "\n";
		}
	}

	return 0;
}

File: Algo-1/week3/5-Bandwidth-Manager/bandwidth_manager.py
# Priority Queue

import heapq


class Packet:
    PROTOCOL_PRIORITY = {
        'ICMP': 10,
        'UDP': 9,
        'RTM': 8,
        'IGMP': 7,
        'DNS': 6,
        'TCP': 5
    }

    def __init__(self, protocol, payload, sequence_number):
        self.priority = Packet.PROTOCOL_PRIORITY[protocol]
        self.payload = payload
        self.sequence_number = sequence_number

    def __lt__(self, other):
        if self.priority == other.priority:
            return self.sequence_number < other.sequence_number
        else:
            return self.priority > other.priority


class BandwidthManager:

    def __init__(self):
        self.heap = []
        self.sequence_number = 0

    # receives a packet with specified protocol and payload
    def rcv(self, protocol, payload):
        new_packet = Packet(protocol, payload, self.sequence_number)
        heapq.heappush(self.heap, new_packet)
        self.sequence_number += 1

    # returns the payload of the packet which should be sent
    def send(self):
        if len(self.heap) == 0:
            return 'Nothing to send!'

        return heapq.heappop(self.heap).payload


def main():
    N = int(input())
    bm = BandwidthManager()

    while N != 0:
        line = input().split()
        if line[0] == 'rcv':
            bm.rcv(line[1], line[2])
        else:
            print(bm.send())

        N -= 1


if __name__ == '__main__':
    main()
